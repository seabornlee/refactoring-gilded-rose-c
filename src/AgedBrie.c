#include <stdlib.h>
#include "AgedBrie.h"

AgedBrie *newAgedBrie(int sellIn, int quality) {
    Item *pItem = newItem("Aged Brie", sellIn, quality);

    AgedBrie *pAgedBrie = (AgedBrie *) malloc(sizeof(AgedBrie));
    if (pAgedBrie == NULL) {
        // printf
        // deleteItem(super)
        return NULL;
    }

    pAgedBrie->base = *pItem;
    pAgedBrie->base.updateQuality = agedBrie_updateQuality;

    return pAgedBrie;
}

void agedBrie_updateQuality(void *pObj) {
    increaseQuality(pObj);
}
