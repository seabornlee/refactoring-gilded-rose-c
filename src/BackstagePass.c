#include "BackstagePass.h"
#include <stdio.h>
#include <stdlib.h>


BackstagePass *newBackstagePass(int sellIn, int quality) {
    BackstagePass *pBackstagePass = (BackstagePass *) malloc(sizeof(BackstagePass));
    if (pBackstagePass == NULL) {
        return NULL;
    }

    Item *pItem = newItem("Backstage passes to a TAFKAL80ETC concert", sellIn, quality);
    pBackstagePass->base = *pItem;

    pBackstagePass->base.updateQuality = backstagePass_updateQuality;
    return pBackstagePass;
}

void backstagePass_updateQuality(void *pObj) {
    Item *item = (Item *) pObj;

    increaseQuality(item);

    if (item->sellIn <= 10) {
        increaseQuality(item);
    }

    if (item->sellIn <= 5) {
        increaseQuality(item);
    }
}
