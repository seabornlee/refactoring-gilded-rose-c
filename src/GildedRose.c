// Copyright 2020 Seaborn Lee
#include <string.h>
#include <stdbool.h>
#include "GildedRose.h"
#include "ItemType.h"

void passOneDay(Item *item);

void updateQuality(Item *item);

void updateSellIn(Item *item);

void updateQualityAfterExpired(Item *item);

void increaseQuality(Item *item);

void decreaseQuality(Item *item);

Item *
init_item(Item *item, const char *name, int sellIn, int quality) {
    item->name = strdup(name);

    if (isSulfuras(item)) {
        item->sellIn = 0;
        item->quality = 80;
    } else {
        if (isAgedBrie(item)) {
            item->sellIn = 0;
        } else {
            item->sellIn = sellIn;
        }
        item->quality = quality;
    }

    return item;
}

void update_quality(Item items[], int size) {
    for (int i = 0; i < size; i++) {
        passOneDay(&items[i]);
    }
}

void passOneDay(Item *item) {
    updateQuality(item);
    updateSellIn(item);

    bool isExpired = item->sellIn < 0;
    if (isExpired) {
        updateQualityAfterExpired(item);
    }
}

void updateQualityAfterExpired(Item *item) {
    if (isAgedBrie(item)) {
        increaseQuality(item);
        return;
    }

    if (isBackstagePass(item)) {
        item->quality = 0;
        return;
    }

    if (!isSulfuras(item)) {
        decreaseQuality(item);
    }
}

void increaseQuality(Item *item) {
    if (item->quality < 50) {
        item->quality = item->quality + 1;
    }
}

void updateSellIn(Item *item) {
    if (isSulfuras(item) || isAgedBrie(item)) return;
    item->sellIn = item->sellIn - 1;
}

void updateQuality(Item *item) {
    if (isAgedBrie(item) || isBackstagePass(item) || isSulfuras(item)) {
        increaseQuality(item);

        if (isBackstagePass(item)) {
            if (item->sellIn < 11) {
                increaseQuality(item);
            }

            if (item->sellIn < 6) {
                increaseQuality(item);
            }
        }
        return;
    }

    decreaseQuality(item);

    if (isManaCake(item)) {
        decreaseQuality(item);
    }
}

void decreaseQuality(Item *item) {
    if (item->quality > 0) {
        item->quality = item->quality - 1;
    }
}