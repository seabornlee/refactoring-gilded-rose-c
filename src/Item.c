#include <string.h>
#include "Item.h"
#include "ItemType.h"

Item *
init_item(Item *item, const char *name, int sellIn, int quality) {
    item->sellIn = sellIn;
    item->quality = quality;
    item->name = strdup(name);

    item->updateQuality = item_updateQuality;

    return item;
}

void updateQualityAfterExpired(Item *item) {
    if (isAgedBrie(item)) {
        increaseQuality(item);
        return;
    }

    if (isBackstagePass(item)) {
        item->quality = 0;
        return;
    }
    if (!isSulfuras(item)) {
        decreaseQuality(item);
    }
}

void item_updateQuality(Item *item) {
    if (!isAgedBrie(item) && !isBackstagePass(item) && !isSulfuras(item)) {
        decreaseQuality(item);
        return;
    }

    increaseQuality(item);
}

void updateSellIn(Item *item) {
    if (isSulfuras(item)) return;
    item->sellIn = item->sellIn - 1;
}

void increaseQuality(Item *item) {
    if (item->quality < 50) {
        item->quality = item->quality + 1;
    }
}

void decreaseQuality(Item *item) {
    if (item->quality > 0) {
        item->quality = item->quality - 1;
    }
}