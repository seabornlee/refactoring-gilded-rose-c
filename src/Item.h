#ifndef REFACTOR_GILDED_ROSE_C_ITEM_H
#define REFACTOR_GILDED_ROSE_C_ITEM_H
#ifdef __cplusplus
extern "C" {
#endif

typedef struct _item Item;

struct _item {
    char *name;
    int sellIn;
    int quality;

    void (* updateQuality)(Item *item);
};


extern Item *init_item(Item *item, const char *name, int sellIn, int quality);
Item *newItem(const char *name, int sellIn, int quality);
void decreaseQuality(Item *item);

void increaseQuality(Item *item);

void updateSellIn(Item *item);

void item_updateQuality(Item *item);

void updateQualityAfterExpired(Item *item);

#ifdef __cplusplus
}

Item *newItem(const char *name, int sellIn, int quality) {
    Item *pItem = (Item *) malloc(sizeof(Item));
    if (pItem == NULL) {
        return NULL;
    }

    pItem->name = strdup(name);
    pItem->sellIn = sellIn;
    pItem->quality = quality;

    pItem->updateQuality= item_updateQuality;

    return pItem;
}

#endif
#endif //REFACTOR_GILDED_ROSE_C_ITEM_H
