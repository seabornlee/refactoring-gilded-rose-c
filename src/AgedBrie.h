#ifndef REFACTOR_GILDED_ROSE_C_AGEDBRIE_H
#define REFACTOR_GILDED_ROSE_C_AGEDBRIE_H

#include "Item.h"
#ifdef __cplusplus
extern "C" {
#endif

typedef struct _aged_brie {
    Item base;
} AgedBrie;

AgedBrie *newAgedBrie(int sellIn, int quality);

static void agedBrie_updateQuality(void *pObj);

#ifdef __cplusplus
}
#endif
#endif //REFACTOR_GILDED_ROSE_C_AGEDBRIE_H
