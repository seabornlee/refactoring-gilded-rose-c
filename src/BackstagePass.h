#ifndef REFACTOR_GILDED_ROSE_C_BACKSTAGEPASS_H
#define REFACTOR_GILDED_ROSE_C_BACKSTAGEPASS_H
#include "Item.h"

#ifdef __cplusplus
extern "C" {
#endif

typedef struct _BackstagePass {
    Item base;
} BackstagePass;

BackstagePass *newBackstagePass(int sellIn, int quality);

static void backstagePass_updateQuality(void *pObj);

#ifdef __cplusplus
}
#endif
#endif //REFACTOR_GILDED_ROSE_C_BACKSTAGEPASS_H
