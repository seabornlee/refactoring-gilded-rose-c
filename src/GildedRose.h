#ifndef REFACTORING_GILDED_ROSE_C_GILDEDROSE_H
#define REFACTORING_GILDED_ROSE_C_GILDEDROSE_H
#include "Item.h"
#ifdef __cplusplus
extern "C" {
#endif

extern void update_quality(Item **items, int size);

#ifdef __cplusplus
}
#endif
#endif //REFACTORING_GILDED_ROSE_C_GILDEDROSE_H
