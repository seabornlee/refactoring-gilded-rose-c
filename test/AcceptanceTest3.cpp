#include <gtest/gtest.h>
#include <ApprovalTests.hpp>
#include <AgedBrie.h>
#include "../src/GildedRose.h"
#include "Item.h"
#include "BackstagePass.h"
#include "Sulfuras.h"


using namespace ApprovalTests;
int
print_item(Item *item) {
    return printf("%s, %d, %d\n", item->name, item->sellIn, item->quality);
}

TEST(AcceptanceTest3, itsLockedDown) {
    testing::internal::CaptureStdout();

    Item *items[9];
    int last = 0;
    int day;
    int index;

    items[last++] = newItem("+5 Dexterity Vest", 10, 20);
//    items[last++] = newItem("Aged Brie", 2, 0);
    items[last++] = (Item *) newAgedBrie(2, 0);
    items[last++] = newItem("Elixir of the Mongoose", 5, 7);
    items[last++] = (Item *) newSulfuras(0, 80);
    items[last++] = (Item *) newSulfuras(-1, 80);
    items[last++] = (Item *) newBackstagePass(15, 20);
    items[last++] = (Item *) newBackstagePass(10, 49);
    items[last++] = (Item *) newBackstagePass(5, 49);

    // this Conjured item doesn't yet work properly
    items[last++] = newItem("Conjured Mana Cake", 3, 6);
//    init_item(items + last++, "Conjured Mana Cake", 3, 6);

    puts("OMGHAI!");

    for (day = 0; day <= 30; day++) {
        printf("-------- day %d --------\n", day);
        printf("name, sellIn, quality\n");
        for (index = 0; index < last; index++) {
            print_item(items[index]);
        }

        printf("\n");

        update_quality(items, last);
    }

    const std::string &out = testing::internal::GetCapturedStdout();
    ApprovalTests::Approvals::verify(out);
}