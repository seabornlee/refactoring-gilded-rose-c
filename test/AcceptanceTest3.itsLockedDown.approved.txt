OMGHAI!
-------- day 0 --------
name, sellIn, quality
+5 Dexterity Vest, 10, 20
Aged Brie, 0, 0
Elixir of the Mongoose, 5, 7
Sulfuras, Hand of Ragnaros, 0, 80
Sulfuras, Hand of Ragnaros, 0, 80
Backstage passes to a TAFKAL80ETC concert, 15, 20
Backstage passes to a TAFKAL80ETC concert, 10, 49
Backstage passes to a TAFKAL80ETC concert, 5, 49
Conjured Mana Cake, 3, 6

-------- day 1 --------
name, sellIn, quality
+5 Dexterity Vest, 9, 19
Aged Brie, 0, 1
Elixir of the Mongoose, 4, 6
Sulfuras, Hand of Ragnaros, 0, 80
Sulfuras, Hand of Ragnaros, 0, 80
Backstage passes to a TAFKAL80ETC concert, 14, 21
Backstage passes to a TAFKAL80ETC concert, 9, 50
Backstage passes to a TAFKAL80ETC concert, 4, 50
Conjured Mana Cake, 2, 4

-------- day 2 --------
name, sellIn, quality
+5 Dexterity Vest, 8, 18
Aged Brie, 0, 2
Elixir of the Mongoose, 3, 5
Sulfuras, Hand of Ragnaros, 0, 80
Sulfuras, Hand of Ragnaros, 0, 80
Backstage passes to a TAFKAL80ETC concert, 13, 22
Backstage passes to a TAFKAL80ETC concert, 8, 50
Backstage passes to a TAFKAL80ETC concert, 3, 50
Conjured Mana Cake, 1, 2

-------- day 3 --------
name, sellIn, quality
+5 Dexterity Vest, 7, 17
Aged Brie, 0, 3
Elixir of the Mongoose, 2, 4
Sulfuras, Hand of Ragnaros, 0, 80
Sulfuras, Hand of Ragnaros, 0, 80
Backstage passes to a TAFKAL80ETC concert, 12, 23
Backstage passes to a TAFKAL80ETC concert, 7, 50
Backstage passes to a TAFKAL80ETC concert, 2, 50
Conjured Mana Cake, 0, 0


